primitives/src/lib.rs

```diff
@@ -553,6 +559,74 @@ impl TryFrom<(&str, AssetIssuer)> for CurrencyId {
 	}
 }
 
+
+pub type ZenlinkAssetId = zenlink_protocol::AssetId;
+const LP_DISCRIMINANT: u64 = 6u64;
+const PARA_CHAIN_ID: u32 = 1000;
+
+impl TryFrom<CurrencyId> for ZenlinkAssetId {
+	type Error = ();
+
+	fn try_from(currency_id: CurrencyId) -> Result<Self, Self::Error> {
+		match currency_id {
+			CurrencyId::Native => Ok(ZenlinkAssetId {
+				chain_id: PARA_CHAIN_ID,
+				asset_type: 0,
+				asset_index: 0 as u64,
+			}),
+			CurrencyId::XCM(symbol) => Ok(ZenlinkAssetId {
+				chain_id: PARA_CHAIN_ID,
+				asset_type: LOCAL,
+				asset_index: symbol as u64,
+			}),
+			CurrencyId::ZenlinkLPToken(symbol0, symbol1) => Ok(ZenlinkAssetId {
+				chain_id: PARA_CHAIN_ID,
+				asset_type: LOCAL,
+				asset_index: (LP_DISCRIMINANT << 8) +
+					((symbol0 as u64 & 0xffff) << 16) +
+					((symbol1 as u64 & 0xffff) << 32),
+			}),
+			_ => Err(()),
+		}
+	}
+}
+
+impl TryFrom<ZenlinkAssetId> for CurrencyId {
+	type Error = ();
+
+	fn try_from(asset_id: ZenlinkAssetId) -> Result<Self, Self::Error> {
+		
+		match asset_id.asset_type {
+			NATIVE => return Ok(CurrencyId::Native),
+			LOCAL => {
+				if asset_id.asset_index & 0x0000_0000_0000_ff00 == 0 {
+					let foreign_id: u8 = asset_id.asset_index.try_into().map_err(|_| {
+						// log::error!("{} has no Foreign Currency Id match.", asset_id.asset_index);
+						()
+					})?;
+
+					return Ok(CurrencyId::XCM(foreign_id));
+				}
+				else {
+					let discriminant = (asset_id.asset_index & 0x0000_0000_0000_ff00) >> 8;
+					return if discriminant == LP_DISCRIMINANT {
+						let token0_id = ((asset_id.asset_index & 0x0000_0000_ffff_0000) >> 16) as u8;
+						let token1_id = ((asset_id.asset_index & 0x0000_ffff_0000_0000) >> 32) as u8; // Should it be >> 32? It was a 16 bits right shift in the example, but i think it was wrong.
```

Comment:
I guess it should actually be `16` but would have to think about the details again. But since they are using that value on their chain in practice I guess we could be quite confident that it works as intended.

Reply:
Thanks for the comment!

I referenced these lines from the [example code](https://github.com/zenlinkpro/Zenlink-DEX-Module/blob/7662acbff71cea3b35db918ceb9370e8929be991/example/primitives/src/lib.rs):

When implementing the [TryFrom trait for ZenlinkAssetId](https://github.com/zenlinkpro/Zenlink-DEX-Module/blob/7662acbff71cea3b35db918ceb9370e8929be991/example/primitives/src/lib.rs#LL107C4-L113C5), they shift the values of the discriminant by 8, symbol0 by 16, and symbol1 by 32. The sum of these values represents the asset_index.

However, during the implementation of the [TryFrom trait for CurrencyId](https://github.com/zenlinkpro/Zenlink-DEX-Module/blob/7662acbff71cea3b35db918ceb9370e8929be991/example/primitives/src/lib.rs#LL128C4-L129C81), they apply the following masks and shifts:

```rust
let token0_id = ((asset_id.asset_index & 0x0000_0000_ffff_0000) >> 16) as u8;
let token1_id = ((asset_id.asset_index & 0x0000_ffff_0000_0000) >> 16) as u8;
```
Based on this code, it appears that token1_id would still be shifted by 16 bits. Additionally, if I'm not mistaken, the result of `(asset_id.asset_index & 0x0000_ffff_0000_0000) >> 16` could potentially exceed the range for u8.

Reply:
Hmm, maybe it's easiest to check which implementation is correct if we just do the 'calculations' based on example values. Could you maybe try both approaches (ie. the one with `>> 16` and the one with `>> 32`)? I mention some example IDs (like an example value for a LP token based on the default generation function) in [this](https://satoshipay.slack.com/archives/C01V1F56RMJ/p1684949005110329) message on slack. So maybe you could try how this turns out when we use `(2124,0,0)`, `(2124,0,1)` for base tokens and assume `(2124,0,4294968832)` is the generated LP token.

Reply:
In terms of the calculation,
asset_id.asset_index is for sure at least a u64 (or this wouldn't compile because the second mask is 64bit)

This should illustrate the behavior this has:
```
    let asset_index = 0x1234_5678_9012_3456_u64;
    let token0_id = (asset_index & 0x0000_0000_ffff_0000) >> 16;
    let token1_id = (asset_index & 0x0000_ffff_0000_0000) >> 16;
    println!("token0_64: {:#x}", token0_id);
    println!("token1_64: {:#x}", token1_id);
    println!("token0_u8: {:#x}", token0_id as u8);
    println!("token1_u8: {:#x}", token1_id as u8);
```
output:
```
token0_64: 0x9012
token1_64: 0x56780000
token0_u8: 0x12
token1_u8: 0x0
```

So, token1_id will always be zero in the end.

So it would make sense if token1 has to be shifted by 32, not 16:
```
    let asset_index = 0x1234_5678_9012_3456_u64;
    let token0_id = (asset_index & 0x0000_0000_ffff_0000) >> 16;
    let token1_id = (asset_index & 0x0000_ffff_0000_0000) >> 32;
    println!("token0_64: {:#x}", token0_id);
    println!("token1_64: {:#x}", token1_id);
    println!("token0_u8: {:#x}", token0_id as u8);
    println!("token1_u8: {:#x}", token1_id as u8);
```
output:
```
token0_64: 0x9012
token1_64: 0x5678
token0_u8: 0x12
token1_u8: 0x78
```

basically, this code is fetching specific bytes inside the asset_index.